docs/writing-tests.md

```diff
@@ -0,0 +1,60 @@
+  - [Good example](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/frontend.feature#L59) - one of the most critical features of GOV.UK ([lower level test exists](https://github.com/alphagov/router/blob/96d5912cec4c8052c2f9f71d257f2add2787b65d/integration_tests/redirect_test.go?L23#L12))
+  - [Bad example](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/collections.feature#L29) - checks an arbitrary, non-critical page ([lower level test exists](https://github.com/alphagov/collections/blob/cc94525a05b2651fee621b588e530aca8e1f77e4/features/viewing_browse.feature))
+  - [Bad example](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/finder_frontend.feature#L45) - missing a lower level test in the app itself ([example test to copy](https://github.com/alphagov/finder-frontend/blob/d70a01903e2813719e7d5adbf79fbcb1d49335fb/spec/controllers/email_alert_subscriptions_controller_spec.rb#L363))
+
+Do not duplicate tests. For example, ["Email signup from an organisation home page"](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/collections.feature#L47) duplicates ["Email signup from a taxon page"](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/collections.feature#L53): both cover data transfer with Email Alert API.
+
+Do not add tests that can be implemented in other ways. We want this to be the tip of the [testing pyramid](https://martinfowler.com/bliki/TestPyramid.html), not a [testing ice cream cone](http://saeedgatson.com/the-software-testing-ice-cream-cone/). Instead of adding a test here, you can:
+
+- Add or extend [health checks for application infrastructure](https://github.com/alphagov/govuk_app_config/blob/main/docs/healthchecks.md). For example, drafting a document in a publishing app involves connecting to infrastructure like a database or AWS S3.
+
+- Add [contract tests](https://docs.publishing.service.gov.uk/manual/pact-broker.html) to cover a chain of APIs. For example, subscribing to an email involves communication with Email Alert API and Content Store, which can be tested via GDS API Adapters.
+
+- Add unit or integration tests to cover in-app behaviour. For example, the way each document format is rendered can be tested in the frontend app that does the rendering.
+
+In contrast, the tests in this repo provide coverage at a higher level. We should be willing to page someone when tests are failing, once they are reliable enough.
+
+## Adding new tests
+
+We use feature files to describe single applications (e.g. [`whitehall`](https://github.com/alphagov/whitehall), [`frontend`](https://github.com/alphagov/frontend)), as well as [cross-application behaviour](https://github.com/alphagov/smokey/blob/main/features/gov_uk.feature).
```

Comment:
Is there a pattern for when you're testing how data flows between applications as to which application you test in? i.e. do you put it in the application you assert against? Only a quibble that if there is a pattern we might want to make that clearer.

Reply:
Yes, we should clarify this. I've already written [the draft guidance for it](https://docs.google.com/document/d/1JvrurIoaYN_Ob_KlY7fecyvG2OGZoZpL0eApeZ5B-14/edit#heading=h.4ia795io3nzc) but have omitted it in this PR because I want to take a similar approach of applying it at the same time as adding it, which will be a lot easier after this change. Right now the tests are all over the place, which is consistent with the (lack of) specific guidance here.

Since we're discussing it, though:

> do you put it in the application you assert against?

Yes, this is what I mean by "target":

> “<app>.feature” should contain tests that target a feature of that app

I've used the word "target" in this guidance for app-to-app data transfer. Do you think it's clear enough, or would "assert about / against" be better?

Reply:
Great stuff. The guidance sounds good. I think target sounds reasonable. Though I interpret it slightly different to assert. I may be caught on an uncommon edge-case though. For example if we're targetting an email-alert-api feature, but we assert on email-alert-frontend I'd assume we'd have an email-alert-api feature to test against?

Reply:
Yup makes sense. I think I'm probably a little too much in edge case land considering how we might be targeting behaviour of an app that is different to what we're asserting against. 

Vague seems reasonable, if possible a slight rephrase to indicate app asserting against could be somewhat beneficial but not worth getting stuck on.

Reply:
Actually, I just found an example:

> “[Check visiting a draft page requires a signon session](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/draft_environment.feature#L9)” should be located in “authenticating_proxy.feature”.

Here we're asserting against government-frontend but actually targeting a feature of the proxy. Another reason to keep it vague is that it would make some of the existing tests appear ineligible, where the transfer of data is only being tested implicitly ([example](https://github.com/alphagov/smokey/blob/68292a3a54475f093f1ff158367cbf5cb94e4e31/features/whitehall.feature#L21)). While we should improve these, I think that should be done as part of a future iteration.

Short version: best to leave it as-is I think - the vagueness is deliberate.